File: Sources/Views/Sidebar/AddProjectPopover.swift
// MARK: - AddProjectPopover
// Popover for the "+" button in the sidebar icon strip.
// Shows recent projects, "Open Folder...", and "Create New..." actions.
// macOS 14+, Swift 5.10

import OSLog
import SwiftUI

// MARK: - AddProjectPopover

struct AddProjectPopover: View {

    @Environment(\.projectManager) private var projectManager
    @Environment(\.dismiss) private var dismiss

    /// Called when the user wants to open a folder via NSOpenPanel.
    let onOpenFolder: () -> Void

    /// Called when the user wants to create a new project. `nil` hides the button.
    let onCreateNew: (() -> Void)?

    @State private var vm: AddProjectViewModel?

    private var viewModel: AddProjectViewModel {
        if let existing = vm { return existing }
        let created = AddProjectViewModel(projectManager: projectManager)
        Task { @MainActor in vm = created }
        return created
    }

    var body: some View {
        let model = viewModel
        VStack(alignment: .leading, spacing: 0) {
            if projectManager.recentHistory.isEmpty {
                emptyState
            } else {
                recentsList(model: model)
            }

            if !projectManager.recentHistory.isEmpty {
                Divider()
                    .overlay(DSColor.borderSubtle)
                    .padding(.vertical, DSSpacing.xs)
            }

            actionButtons
        }
        .padding(.horizontal, DSSpacing.md)
        .padding(.vertical, DSSpacing.sm)
        .frame(minWidth: 280, idealWidth: DSLayout.addProjectPopoverWidth, maxWidth: 360)
        .background(DSColor.surfaceOverlay)
        .onAppear {
            if vm == nil {
                vm = AddProjectViewModel(projectManager: projectManager)
            }
        }
    }

    // MARK: - Empty State

    @ViewBuilder
    private var emptyState: some View {
        VStack(spacing: DSSpacing.sm) {
            Image(systemName: "folder.badge.questionmark")
                .font(DSFont.emptyStateIcon)
                .foregroundStyle(DSColor.textMuted)

            Text("No recent projects")
                .font(DSFont.sidebarItem)
                .foregroundStyle(DSColor.textSecondary)

            Text("Open a folder or create a new project")
                .font(DSFont.sidebarItemSmall)
                .foregroundStyle(DSColor.textMuted)
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, DSSpacing.md)

        Divider()
            .overlay(DSColor.borderSubtle)
            .padding(.vertical, DSSpacing.xs)
    }

    // MARK: - Recents List

    private func recentsList(model: AddProjectViewModel) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            Text("RECENT")
                .font(DSFont.sidebarSection)
                .foregroundStyle(DSColor.textSecondary)
                .padding(.bottom, DSSpacing.xs)

            ForEach(projectManager.recentHistory) { project in
                ProjectListItemView(project: project, onTap: {
                    if model.openRecentProject(project) {
                        dismiss()
                    }
                }) {
                    Text(project.lastOpened, format: .relative(presentation: .named))
                        .font(DSFont.sidebarItemSmall)
                        .foregroundStyle(DSColor.textSecondary)
                        .lineLimit(1)
                }
            }

            if let err = model.openError {
                Text(err)
                    .font(DSFont.sidebarItemSmall)
                    .foregroundStyle(DSColor.gitDeleted)
                    .lineLimit(2)
                    .padding(.top, DSSpacing.xs)
            }
        }
    }

    // MARK: - Action Buttons

    private var actionButtons: some View {
        VStack(alignment: .leading, spacing: 0) {
            ActionRow(
                title: "Open Folder...",
                systemImage: "folder.badge.plus"
            ) {
                onOpenFolder()
                dismiss()
            }

            if let onCreateNew {
                ActionRow(
                    title: "Create New...",
                    systemImage: "plus.square"
                ) {
                    onCreateNew()
                    dismiss()
                }
            }
        }
    }
}

// MARK: - ActionRow

/// A styled action button row (e.g. "Open Folder...", "Create New...").
private struct ActionRow: View {

    let title: String
    let systemImage: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Label(title, systemImage: systemImage)
                .font(DSFont.sidebarItem)
                .foregroundStyle(DSColor.textPrimary)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, DSSpacing.md)
                .padding(.vertical, DSSpacing.sm)
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .sidebarHover(cornerRadius: DSRadius.sm)
    }
}

File: desktopApp/src/main/kotlin/studio/vibe/desktop/ui/AddProjectPopover.kt
@file:OptIn(kotlin.uuid.ExperimentalUuidApi::class)

package studio.vibe.desktop.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.hoverable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsHoveredAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Folder
import androidx.compose.material.icons.filled.FolderOpen
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import studio.vibe.shared.contract.ProjectManaging
import studio.vibe.desktop.ui.theme.DSColor
import studio.vibe.desktop.ui.theme.LocalDSColors
import studio.vibe.desktop.ui.theme.DSFont
import studio.vibe.desktop.ui.theme.DSRadius
import studio.vibe.desktop.ui.theme.DSSpacing
import studio.vibe.shared.model.FilePath
import studio.vibe.shared.model.Project
import studio.vibe.shared.viewmodel.AddProjectViewModel
import kotlin.time.ExperimentalTime
import kotlin.time.Instant
import kotlin.time.Clock

// ── AddProjectPopover ─────────────────────────────────────────────────────────

/**
 * Popup/dropdown for the "+" button in the sidebar icon strip.
 *
 * Mirrors AddProjectPopover.swift (225 LOC).
 *
 * Shows:
 * - Recent projects list (folder icon, name, path, relative date)
 * - Empty state when no recent projects
 * - "Open Folder..." action button
 * - Optional "Create New..." action button
 *
 * Width: 280–360dp.
 *
 * @param container DI container providing [studio.vibe.shared.contract.ProjectManaging].
 * @param onOpenFolder Called when user taps "Open Folder...".
 * @param onCreateNew Called when user taps "Create New..."; pass null to hide the button.
 * @param onDismiss Called after a project is opened or actions are triggered.
 */
@Composable
public fun AddProjectPopover(
    projectStore: ProjectManaging,
    onOpenFolder: () -> Unit,
    onCreateNew: (() -> Unit)? = null,
    onDismiss: () -> Unit,
) {
    val coroutineScope = rememberCoroutineScope()
    val vm = remember {
        AddProjectViewModel(
            projectManaging = projectStore,
            scope = coroutineScope,
        )
    }
    val state by vm.state.collectAsState()
    val recentProjects = state.recentProjects

    Column(
        modifier = Modifier
            .widthIn(min = 280.dp, max = 360.dp)
            .background(LocalDSColors.current.surfaceOverlay)
            .padding(horizontal = DSSpacing.md, vertical = DSSpacing.sm),
    ) {
        if (recentProjects.isEmpty()) {
            // ── Empty state ───────────────────────────────────────────────
            PopoverEmptyState()
            HorizontalDivider(
                color = LocalDSColors.current.borderSubtle,
                thickness = 1.dp,
                modifier = Modifier.padding(vertical = DSSpacing.xs),
            )
        } else {
            // ── Recents list ──────────────────────────────────────────────
            Text(
                text = "RECENT",
                style = DSFont.sidebarSection,
                color = LocalDSColors.current.textSecondary,
                modifier = Modifier.padding(bottom = DSSpacing.xs),
            )
            recentProjects.forEach { project ->
                RecentRow(
                    project = project,
                    onTap = {
                        vm.openRecentProject(project.path)
                        onDismiss()
                    },
                )
            }
            if (state.errorMessage != null) {
                Text(
                    text = state.errorMessage!!,
                    style = DSFont.sidebarItemSmall,
                    color = LocalDSColors.current.gitDeleted,
                    maxLines = 2,
                    modifier = Modifier.padding(top = DSSpacing.xs),
                )
            }
            HorizontalDivider(
                color = LocalDSColors.current.borderSubtle,
                thickness = 1.dp,
                modifier = Modifier.padding(vertical = DSSpacing.xs),
            )
        }

        // ── Action buttons ────────────────────────────────────────────────
        ActionRow(
            title = "Open Folder...",
            icon = Icons.Filled.FolderOpen,
        ) {
            onOpenFolder()
            onDismiss()
        }

        if (onCreateNew != null) {
            ActionRow(
                title = "Create New...",
                icon = Icons.Filled.Add,
            ) {
                onCreateNew()
                onDismiss()
            }
        }
    }
}

// ── EmptyState ────────────────────────────────────────────────────────────────

@Composable
private fun PopoverEmptyState() {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = DSSpacing.md),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(DSSpacing.sm),
    ) {
        Icon(
            imageVector = Icons.Filled.Folder,
            contentDescription = null,
            tint = LocalDSColors.current.textMuted,
            modifier = Modifier.size(28.dp),
        )
        Text(
            text = "No recent projects",
            style = DSFont.sidebarItem,
            color = LocalDSColors.current.textSecondary,
        )
        Text(
            text = "Open a folder or create a new project",
            style = DSFont.sidebarItemSmall,
            color = LocalDSColors.current.textMuted,
        )
    }
}

// ── RecentRow ─────────────────────────────────────────────────────────────────

/**
 * Single row displaying a recent project: folder icon, name, tilde-abbreviated path,
 * and a relative date label.
 */
@Composable
private fun RecentRow(
    project: Project,
    onTap: () -> Unit,
) {
    val interactionSource = remember { MutableInteractionSource() }
    val isHovered by interactionSource.collectIsHoveredAsState()

    Box(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(DSRadius.sm))
            .background(if (isHovered) LocalDSColors.current.hoverOverlay else LocalDSColors.current.surfaceOverlay.copy(alpha = 0f))
            .hoverable(interactionSource)
            .clickable(onClick = onTap)
            .padding(horizontal = DSSpacing.sm, vertical = DSSpacing.xs),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                imageVector = Icons.Filled.Folder,
                contentDescription = null,
                tint = LocalDSColors.current.gitModified,
                modifier = Modifier.size(16.dp),
            )
            Spacer(Modifier.width(DSSpacing.sm))
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = project.name,
                    style = DSFont.sidebarItem,
                    color = LocalDSColors.current.textPrimary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        text = project.path.path.replace(System.getProperty("user.home") ?: "", "~"),
                        style = DSFont.sidebarItemSmall,
                        color = LocalDSColors.current.textMuted,
                        maxLines = 1,
                        overflow = TextOverflow.MiddleEllipsis,
                        modifier = Modifier.weight(1f),
                    )
                    Spacer(Modifier.width(DSSpacing.xs))
                    Text(
                        text = formatRelativeDate(project.lastOpened),
                        style = DSFont.sidebarItemSmall,
                        color = LocalDSColors.current.textSecondary,
                        maxLines = 1,
                    )
                }
            }
        }
    }
}

// ── ActionRow ─────────────────────────────────────────────────────────────────

/**
 * Styled action button row (e.g. "Open Folder...", "Create New...").
 */
@Composable
private fun ActionRow(
    title: String,
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    action: () -> Unit,
) {
    val interactionSource = remember { MutableInteractionSource() }
    val isHovered by interactionSource.collectIsHoveredAsState()

    Box(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(DSRadius.sm))
            .background(if (isHovered) LocalDSColors.current.hoverOverlay else LocalDSColors.current.surfaceOverlay.copy(alpha = 0f))
            .hoverable(interactionSource)
            .clickable(onClick = action)
            .padding(horizontal = DSSpacing.md, vertical = DSSpacing.sm),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = LocalDSColors.current.textPrimary,
                modifier = Modifier.size(14.dp),
            )
            Spacer(Modifier.width(DSSpacing.sm))
            Text(
                text = title,
                style = DSFont.sidebarItem,
                color = LocalDSColors.current.textPrimary,
            )
        }
    }
}

// ── Helpers ───────────────────────────────────────────────────────────────────

/**
 * Formats an [Instant] as a human-readable relative date string.
 *
 * Mirrors Swift's `.relative(presentation: .named)` format style.
 */
@OptIn(ExperimentalTime::class)
private fun formatRelativeDate(instant: Instant): String {
    val now = Clock.System.now()
    val delta = now - instant
    val totalSeconds = delta.inWholeSeconds
    return when {
        totalSeconds < 60 -> "только что"
        totalSeconds < 3600 -> "${totalSeconds / 60} мин. назад"
        totalSeconds < 86400 -> "${totalSeconds / 3600} ч. назад"
        totalSeconds < 86400 * 7 -> "${totalSeconds / 86400} д. назад"
        totalSeconds < 86400 * 30 -> "${totalSeconds / (86400 * 7)} нед. назад"
        totalSeconds < 86400 * 365 -> "${totalSeconds / (86400 * 30)} мес. назад"
        else -> "${totalSeconds / (86400 * 365)} г. назад"
    }
}
